### PR TITLE
Add Lorentzian solution to the Poisson equation

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

- A 3D spherically symmetric solution with a Lorentzian profile:
bildschirmfoto 2019-01-09 um 12 15 32
- This solution falls off with the typical `1/r` dependence, which makes it useful to test spherical domains
- It could be generalized to 1D and 2D, e.g. with solutions that scale as `r` and `log(r)`, respectively. But I'm not sure if we'll actually need those so I'd like to defer that but leave the option open.
- I sneaked in another commit that fixes the normalization in the Poisson observe action

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
@moxcodes @wthrowe Github suggested you as reviewers, so I added you both, trying out the new code review guidelines :) I realise we have not put those guidelines into effect yet, but thought we could do some trial runs. So if you don't have time to review this PR in the next two days, please feel free to forward to another reviewer. If you do agree to review this PR, I think the idea is that once the two of you are happy with this PR (potentially after a few rounds of review) it may be merged by @wthrowe directly (since he has merge rights), or forwarded to @sxs-collaboration/spectre-core-devs for a brief final review and then merged. Does this sound about right to you? Since this PR is fairly straight forward to review, it seems like a good candidate for a trial run :)

**Comment by moxcodes** [comment]
I'm happy to review.

File: src/Elliptic/Initialization/LinearSolver.hpp
Comment (nilsvu) [inline src/Elliptic/Initialization/LinearSolver.hpp]: Oh this change shouldn’t be here, it accidentally made its way into this PR during a rebase. Will remove asap.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp:37]: I don't really see a need to template this.  If you want to I'm OK with it, but add a static_assert that Dim is 3.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: Move to cpp.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: Move to cpp.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp]: You don't use this.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Travis is consistently failing with an ICE.  The tests don't pass.

I believe you have correctly summarized the new code review policy that has not yet been formally proposed.  Since it has not been approved by the collaboration, I'm not going to use it as a justification to ignore the established merge procedure.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp]: can this one be const too?

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: would there be any gain in making these static? This solution at least doesn't seem to depend on any stored parameters.

**Review by moxcodes — COMMENTED** [review]
Apart from the compiler error noted by @wthrowe, I only have minor non-blocking comments to make.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: I though with the `constexpr` this is would be inlined, so it would trivially evaluate to `true`, which would be better than moving this to the cpp. Am I wrong, and if so, why? I made the change in the fixup commit, just would like to understand what's going on.

**Comment by nilsvu** [comment]
@moxcodes @wthrowe thanks for reviewing, I added a fixup with your requested changes. Also added another fixup that brings the way the variables are retrieved in line with how it's done in the evolution code, i.e. a single `variables` function that has an overload for every variable and a collection of variables. I move the other elliptic analytic solutions to the architecture in #1266.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: That could work, it's in the last fixup. @wthrowe do you know of any subtleties that would make this being `static` incompatible to the non-static interface of the other analytic solutions?

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: Fore rarely called functions (this one will probably never be called outside the test) we would prefer to take the slight runtime performance hit to get the slight compile-time performance improvement from not having to parse it when the hpp is included.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.hpp]: I didn't realize it was valid C++ to use static functions like that, but it is.  I don't know if there's any gain, but I don't see any harm to it either so I'm fine with the change.

**Comment by wthrowe** [comment]
I've replied to the comment threads.  I don't have time to look at the code changes right now, but I'll get to that later today.  Travis is showing linking failures, probably from a missing instantiation.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp
Comment (moxcodes) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp:18]: looks like the template syntax didn't go away in the cpp when it was updated in the hpp.

**Review by moxcodes — CHANGES_REQUESTED** [review]
I'm happy with the changes, just need to finish the template removal and appease IWYU.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp:18]: This code looks correct to me.  None of the template structure has changed here.  (Just a rename of the function.)

**Comment by wthrowe** [comment]
I'm happy with the fixups.  Just make Travis happy too.

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp
Comment (moxcodes) [reply src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp:18]: agreed, that's my mistake. I read the fixups too quickly and thought the template on dimension was removed.

**Comment by nilsvu** [comment]
done @moxcodes @wthrowe. This should be ready @sxs-collaboration/spectre-core-devs

**Review by wthrowe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
